File: examples/csharp/streaming/task-events/Agent.cs
using System.Text.Json;
using System.Text.Json.Nodes;

public static class Agent
{
    private const string KubeMqUrl = "http://localhost:9090";
    private const string AgentId = "task-events-agent-01";
    private const int AgentPort = 18080;

    public static async Task RunAsync()
    {
        var builder = WebApplication.CreateBuilder();
        builder.Logging.ClearProviders();
        var app = builder.Build();

        app.MapPost("/", async (HttpContext context) =>
        {
            var body = (await JsonSerializer.DeserializeAsync<JsonNode>(context.Request.Body))!;
            var method = body["method"]?.GetValue<string>() ?? "";

            if (method == "message/stream")
            {
                context.Response.ContentType = "text/event-stream";
                context.Response.Headers.CacheControl = "no-cache";

                var status1 = JsonSerializer.Serialize(new { type = "status_update", payload = new { status = "working", progress = 1, total = 3 } });
                await context.Response.WriteAsync($"event: task.status\ndata: {status1}\n\n");
                await context.Response.Body.FlushAsync();
                await Task.Delay(500);

                var status2 = JsonSerializer.Serialize(new { type = "status_update", payload = new { status = "working", progress = 2, total = 3 } });
                await context.Response.WriteAsync($"event: task.status\ndata: {status2}\n\n");
                await context.Response.Body.FlushAsync();
                await Task.Delay(500);

                var artifact = JsonSerializer.Serialize(new { type = "artifact", payload = new { name = "result.json", data = new { key = "value" } } });
                await context.Response.WriteAsync($"event: task.artifact\ndata: {artifact}\n\n");
                await context.Response.Body.FlushAsync();
                await Task.Delay(300);

                var done = JsonSerializer.Serialize(new { type = "done", payload = new { final_result = "completed" } });
                await context.Response.WriteAsync($"event: task.done\ndata: {done}\n\n");
                await context.Response.Body.FlushAsync();
                return;
            }

            var response = new JsonObject
            {
                ["jsonrpc"] = "2.0",
                ["id"] = body["id"]?.DeepClone(),
                ["result"] = new JsonObject { ["echo"] = body.DeepClone() }
            };
            context.Response.ContentType = "application/json";
            await context.Response.WriteAsync(response.ToJsonString());
        });

        app.Urls.Add($"http://0.0.0.0:{AgentPort}");
        await app.StartAsync();
        Console.WriteLine($"Agent listening on port {AgentPort}");

        await RegisterAgentAsync();

        await app.WaitForShutdownAsync();
    }

    private static async Task RegisterAgentAsync()
    {
        var card = new JsonObject
        {
            ["agent_id"] = AgentId,
            ["name"] = "Task Events Agent",
            ["description"] = "Emits various event types",
            ["version"] = "1.0.0",
            ["url"] = $"http://localhost:{AgentPort}/",
            ["skills"] = new JsonArray(),
            ["defaultInputModes"] = new JsonArray("text"),
            ["defaultOutputModes"] = new JsonArray("text"),
            ["protocolVersions"] = new JsonArray("1.0")
        };

        using var client = new HttpClient();
        var resp = await client.PostAsync(
            $"{KubeMqUrl}/agents/register",
            new StringContent(card.ToJsonString(), System.Text.Encoding.UTF8, "application/json"));
        Console.WriteLine($"Registered: {(int)resp.StatusCode}");
    }
}

File: examples/java/streaming/task-events/Agent.java
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;

import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;

public class Agent {

    static final String KUBEMQ_URL = "http://localhost:9090";
    static final String AGENT_ID = "task-events-agent-01";
    static final int AGENT_PORT = 18080;
    static final ObjectMapper MAPPER = new ObjectMapper();

    public static void main(String[] args) throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress(AGENT_PORT), 0);
        server.createContext("/", exchange -> {
            if (!"POST".equals(exchange.getRequestMethod())) {
                exchange.sendResponseHeaders(405, -1);
                return;
            }
            var body = MAPPER.readTree(exchange.getRequestBody());
            var method = body.path("method").asText("");

            if ("message/stream".equals(method)) {
                exchange.getResponseHeaders().set("Content-Type", "text/event-stream");
                exchange.getResponseHeaders().set("Cache-Control", "no-cache");
                exchange.sendResponseHeaders(200, 0);
                try (OutputStream os = exchange.getResponseBody()) {
                    var s1 = MAPPER.writeValueAsString(Map.of(
                        "type", "status_update",
                        "payload", Map.of("status", "working", "progress", 1, "total", 3)));
                    os.write(("event: task.status\ndata: " + s1 + "\n\n").getBytes());
                    os.flush();
                    Thread.sleep(500);

                    var s2 = MAPPER.writeValueAsString(Map.of(
                        "type", "status_update",
                        "payload", Map.of("status", "working", "progress", 2, "total", 3)));
                    os.write(("event: task.status\ndata: " + s2 + "\n\n").getBytes());
                    os.flush();
                    Thread.sleep(500);

                    var artifact = MAPPER.writeValueAsString(Map.of(
                        "type", "artifact",
                        "payload", Map.of("name", "result.json", "data", Map.of("key", "value"))));
                    os.write(("event: task.artifact\ndata: " + artifact + "\n\n").getBytes());
                    os.flush();
                    Thread.sleep(300);

                    var done = MAPPER.writeValueAsString(Map.of(
                        "type", "done",
                        "payload", Map.of("final_result", "completed")));
                    os.write(("event: task.done\ndata: " + done + "\n\n").getBytes());
                    os.flush();
                }
                return;
            }

            var response = MAPPER.createObjectNode();
            response.put("jsonrpc", "2.0");
            response.set("id", body.get("id"));
            response.putObject("result").set("echo", body);
            byte[] out = MAPPER.writeValueAsBytes(response);
            exchange.getResponseHeaders().set("Content-Type", "application/json");
            exchange.sendResponseHeaders(200, out.length);
            try (OutputStream os = exchange.getResponseBody()) { os.write(out); }
        });
        server.start();
        System.out.println("Agent listening on port " + AGENT_PORT);

        var card = Map.of(
            "agent_id", AGENT_ID,
            "name", "Task Events Agent",
            "description", "Emits various event types",
            "version", "1.0.0",
            "url", "http://localhost:" + AGENT_PORT + "/",
            "skills", List.of(),
            "defaultInputModes", List.of("text"),
            "defaultOutputModes", List.of("text"),
            "protocolVersions", List.of("1.0")
        );

        var client = HttpClient.newHttpClient();
        var req = HttpRequest.newBuilder()
            .uri(URI.create(KUBEMQ_URL + "/agents/register"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(card)))
            .build();
        var resp = client.send(req, HttpResponse.BodyHandlers.ofString());
        System.out.println("Registered: " + resp.statusCode());

        Thread.currentThread().join();
    }
}
